Update sql db & elastic-pool to support sku model

jaredmoo opened the pull request:
This is a major change to the db & elastic-pool types, so I am taking care to make sure that there is no breaking change to command input parameters, and I am using output transformers (where possible, since transforms do not work for long-running operations) to populate important properties of return objects. I have also added table formatters to db and elastic-pool commands.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

---

Additional checklist:

- [x] Add e2e tests for new scenarios (create with family/capacity, update from Standard to GP and back, etc)
- [x] Add help and examples for new scenarios

promptws commented:
View a preview at https://prompt.ws/r/Azure/azure-cli/6286
This is an **experimental preview** for @microsoft users.

jaredmoo commented:
# Syntax changes (Part 1: db create & update)

## az sql db copy

### Before

```
Command
    az sql db copy: Create a copy of a database.

Arguments
    --dest-name [Required]: Name of the database that will be created as the copy destination.
    --dest-resource-group : Name of the resouce group to create the copy in. If unspecified,
                            defaults to the origin resource group.
    --dest-server         : Name of the server to create the copy in. If unspecified, defaults to
                            the origin server.
    --elastic-pool        : Name of the elastic pool to create the new database in.
    --no-wait             : Do not wait for the long-running operation to finish.
    --service-objective   : Name of the service objective for the new database.
    --tags                : Resource tags.

Resource Id Arguments
    --ids                 : One or more resource IDs (space-delimited). If provided, no other
                            'Resource Id' arguments should be specified.
    --name -n             : Name of the Azure SQL Database.
    --resource-group -g   : Name of resource group. You can configure the default group using `az
                            configure --defaults group=<name>`.
    --server -s           : Name of the Azure SQL server.
```

### After

```
Command
    az sql db copy: Create a copy of a database.
        A full list of performance level options can be seen by executing `az sql db list-editions
        -a -o table -l LOCATION`. The copy destination database must have the same edition as the
        source database, but you can change the edition after the copy has completed.

Arguments
    --dest-name [Required]: Name of the database that will be created as the copy destination.
    --dest-resource-group : Name of the resouce group to create the copy in. If unspecified,
                            defaults to the origin resource group.
    --dest-server         : Name of the server to create the copy in. If unspecified, defaults to
                            the origin server.
    --license-type        : The license type to apply for this database.  Allowed values: BasePrice,
                            LicenseIncluded.
    --no-wait             : Do not wait for the long-running operation to finish.
    --tags                : Resource tags.
    --zone-redundant -z   : Specifies whether to enable zone redundancy.  Allowed values: false,
                            true.

Performance Level (components) Arguments
    --capacity -c         : The capacity component of the sku in integer number of DTUs or vcores.
    --family -f           : The compute generation component of the sku (for vcore skus only).
                            Allowed values include: Gen4, Gen5.

Performance Level Arguments
    --elastic-pool        : The name or resource id of the elastic pool to create the database in.
    --service-objective   : The service objective for the new database. For example: Basic, S0, P1,
                            GP_Gen4_1, BC_Gen5_2.

Resource Id Arguments
    --ids                 : One or more resource IDs (space-delimited). If provided, no other
                            'Resource Id' arguments should be specified.
    --name -n             : Name of the Azure SQL Database.
    --resource-group -g   : Name of resource group. You can configure the default group using `az
                            configure --defaults group=<name>`.
    --server -s           : Name of the Azure SQL server.
```

## az sql db create

### Before

```
Command
    az sql db create: Create a database.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --collation                   : The collation of the database. If createMode is not Default,
                                    this value is ignored.
    --edition                     : The edition of the database.
    --elastic-pool                : The name of the elastic pool the database is in. If
                                    elasticPoolName and requestedServiceObjectiveName are both
                                    updated, the value of requestedServiceObjectiveName is ignored.
                                    Not supported for DataWarehouse edition.
    --max-size                    : The max storage size of the database. Only the following sizes
                                    are supported (in addition to limitations being placed on each
                                    edition): 100MB, 500MB, 1GB, 5GB, 10GB, 20GB, 30GB, 150GB,
                                    200GB, 500GB. If no unit is specified, defaults to bytes (B).
    --no-wait                     : Do not wait for the long-running operation to finish.
    --sample-name                 : Indicates the name of the sample schema to apply when creating
                                    this database. If createMode is not Default, this value is
                                    ignored. Not supported for DataWarehouse edition.
    --service-objective           : The name of the configured service level objective of the
                                    database. This is the service level objective that is in the
                                    process of being applied to the database. Once successfully
                                    updated, it will match the value of serviceLevelObjective
                                    property. To see possible values, query the capabilities API (/s
                                    ubscriptions/{subscriptionId}/providers/Microsoft.Sql/locations/
                                    {locationID}/capabilities) referred to by operationId:
                                    "Capabilities_ListByLocation." or use the Azure CLI command `az
                                    sql db list-editions --location <location> --query
                                    [].supportedServiceLevelObjectives[].name`.
    --tags                        : Resource tags.
    --zone-redundant -z           : Specifies whether to enable zone redundancy for the database.
                                    Allowed values: false, true.
```

### After

```
Command
    az sql db create: Create a database.
        A full list of performance level options can be seen by executing `az sql db list-editions
        -a -o table -l LOCATION`.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --license-type                : The license type to apply for this database.  Allowed values:
                                    BasePrice, LicenseIncluded.
    --max-size                    : The max storage size. If no unit is specified, defaults to bytes
                                    (B).
    --no-wait                     : Do not wait for the long-running operation to finish.
    --tags                        : Resource tags.
    --zone-redundant -z           : Specifies whether to enable zone redundancy.  Allowed values:
                                    false, true.

Creation Arguments
    --catalog-collation           : Collation of the metadata catalog.  Allowed values:
                                    DATABASE_DEFAULT, SQL_Latin1_General_CP1_CI_AS.
    --collation                   : The collation of the database.
    --sample-name                 : The name of the sample schema to apply when creating this
                                    database.

Performance Level (components) Arguments
    --capacity -c                 : The capacity component of the sku in integer number of DTUs or
                                    vcores.
    --edition --tier -e           : The edition component of the sku. Allowed values include: Basic,
                                    Standard, Premium, GeneralPurpose, BusinessCritical.
    --family -f                   : The compute generation component of the sku (for vcore skus
                                    only). Allowed values include: Gen4, Gen5.

Performance Level Arguments
    --elastic-pool                : The name or resource id of the elastic pool to create the
                                    database in.
    --service-objective           : The service objective for the new database. For example: Basic,
                                    S0, P1, GP_Gen4_1, BC_Gen5_2.
```

## az sql db restore

### Before

```
ommand
    az sql db restore: Create a new database by restoring from a backup.

Arguments
    --dest-name [Required]: Name of the database that will be created as the restore destination.
    --edition             : The edition for the new database.
    --elastic-pool        : Name of the elastic pool to create the new database in.
    --no-wait             : Do not wait for the long-running operation to finish.
    --service-objective   : Name of service objective for the new database.
    --tags                : Resource tags.

Resource Id Arguments
    --ids                 : One or more resource IDs (space-delimited). If provided, no other
                            'Resource Id' arguments should be specified.
    --name -n             : Name of the Azure SQL Database.
    --resource-group -g   : Name of resource group. You can configure the default group using `az
                            configure --defaults group=<name>`.
    --server -s           : Name of the Azure SQL server.

Restore Point Arguments
    --deleted-time        : If specified, restore from a deleted database instead of from an
                            existing database. Must match the deleted time of a deleted database in
                            the same server. Either --time or --deleted-time (or both) must be
                            specified.
    --time -t             : The point in time of the source database that will be restored to create
                            the new database. Must be greater than or equal to the source database's
                            earliestRestoreDate value. Either --time or --deleted-time (or both)
                            must be specified.
```

### After

```
Command
    az sql db restore: Create a new database by restoring from a backup.
        A full list of performance level options can be seen by executing `az sql db list-editions
        -a -o table -l LOCATION`.

Arguments
    --dest-name [Required]: Name of the database that will be created as the restore destination.
    --license-type        : The license type to apply for this database.  Allowed values: BasePrice,
                            LicenseIncluded.
    --no-wait             : Do not wait for the long-running operation to finish.
    --tags                : Resource tags.
    --zone-redundant -z   : Specifies whether to enable zone redundancy.  Allowed values: false,
                            true.

Performance Level (components) Arguments
    --capacity -c         : The capacity component of the sku in integer number of DTUs or vcores.
    --edition --tier -e   : The edition component of the sku. Allowed values include: Basic,
                            Standard, Premium, GeneralPurpose, BusinessCritical.
    --family -f           : The compute generation component of the sku (for vcore skus only).
                            Allowed values include: Gen4, Gen5.

Performance Level Arguments
    --elastic-pool        : The name or resource id of the elastic pool to create the database in.
    --service-objective   : The service objective for the new database. For example: Basic, S0, P1,
                            GP_Gen4_1, BC_Gen5_2.

Resource Id Arguments
    --ids                 : One or more resource IDs (space-delimited). If provided, no other
                            'Resource Id' arguments should be specified.
    --name -n             : Name of the Azure SQL Database.
    --resource-group -g   : Name of resource group. You can configure the default group using `az
                            configure --defaults group=<name>`.
    --server -s           : Name of the Azure SQL server.

Restore Point Arguments
    --deleted-time        : If specified, restore from a deleted database instead of from an
                            existing database. Must match the deleted time of a deleted database in
                            the same server. Either --time or --deleted-time (or both) must be
                            specified.
    --time -t             : The point in time of the source database that will be restored to create
                            the new database. Must be greater than or equal to the source database's
                            earliestRestoreDate value. Either --time or --deleted-time (or both)
                            must be specified.
```

## az sql db replica create

### Before

```
Command
    az sql db replica create: Create a database as a readable secondary replica of an existing
    database.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --partner-server    [Required]: Name of the server to create the new replica in.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --elastic-pool                : Name of elastic pool to create the new replica in.
    --no-wait                     : Do not wait for the long-running operation to finish.
    --partner-resource-group      : Name of the resource group to create the new replica in. If
                                    unspecified, defaults to the origin resource group.
    --service-objective           : Name of service objective for the new replica.
    --tags                        : Resource tags.
```

### After

```
Command
    az sql db replica create: Create a database as a readable secondary replica of an existing
    database.
        A full list of performance level options can be seen by executing `az sql db list-editions
        -a -o table -l LOCATION`. The secondary database must have the same edition as the primary
        database.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --partner-server    [Required]: Name of the server to create the new replica in.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --license-type                : The license type to apply for this database.  Allowed values:
                                    BasePrice, LicenseIncluded.
    --no-wait                     : Do not wait for the long-running operation to finish.
    --partner-resource-group      : Name of the resource group to create the new replica in. If
                                    unspecified, defaults to the origin resource group.
    --tags                        : Resource tags.
    --zone-redundant -z           : Specifies whether to enable zone redundancy.  Allowed values:
                                    false, true.

Performance Level (components) Arguments
    --capacity -c                 : The capacity component of the sku in integer number of DTUs or
                                    vcores.
    --family -f                   : The compute generation component of the sku (for vcore skus
                                    only). Allowed values include: Gen4, Gen5.

Performance Level Arguments
    --elastic-pool                : The name or resource id of the elastic pool to create the
                                    database in.
    --service-objective           : The service objective for the new database. For example: Basic,
                                    S0, P1, GP_Gen4_1, BC_Gen5_2.
```

## az sql db update

### Before

```
Command
    az sql db update: Update a database.

Arguments
    --elastic-pool     : The name of the elastic pool to move the database into.
    --max-size         : The new maximum size of the database expressed in bytes.
    --no-wait          : Do not wait for the long-running operation to finish.
    --service-objective: The name of the new service objective. If this is a standalone db service
                         objective and the db is currently in an elastic pool, then the db is
                         removed from the pool.
    --zone-redundant -z: Specifies whether to enable zone redundancy for the database.  Allowed
                         values: false, true.

Generic Update Arguments
    --add              : Add an object to a list of objects by specifying a path and key value
                         pairs.  Example: --add property.listProperty <key=value, string or JSON
                         string>.
    --remove           : Remove a property or an element from a list.  Example: --remove
                         property.list <indexToRemove> OR --remove propertyToRemove.
    --set              : Update an object by specifying a property path and value to set.  Example:
                         --set property1.property2=<value>.

Resource Id Arguments
    --ids              : One or more resource IDs (space-delimited). If provided, no other 'Resource
                         Id' arguments should be specified.
    --name -n          : Name of the Azure SQL Database.
    --resource-group -g: Name of resource group. You can configure the default group using `az
                         configure --defaults group=<name>`.
    --server -s        : Name of the Azure SQL server.
```

### After

```
Command
    az sql db update: Update a database.

Arguments
    --max-size         : The new maximum size of the database expressed in bytes.
    --no-wait          : Do not wait for the long-running operation to finish.
    --zone-redundant -z: Specifies whether to enable zone redundancy.  Allowed values: false, true.

Generic Update Arguments
    --add              : Add an object to a list of objects by specifying a path and key value
                         pairs.  Example: --add property.listProperty <key=value, string or JSON
                         string>.
    --remove           : Remove a property or an element from a list.  Example: --remove
                         property.list <indexToRemove> OR --remove propertyToRemove.
    --set              : Update an object by specifying a property path and value to set.  Example:
                         --set property1.property2=<value>.

Performance Level (components) Arguments
    --capacity -c      : The capacity component of the sku in integer number of DTUs or vcores.
    --edition --tier -e: The edition component of the sku. Allowed values include: Basic, Standard,
                         Premium, GeneralPurpose, BusinessCritical.
    --family -f        : The compute generation component of the sku (for vcore skus only). Allowed
                         values include: Gen4, Gen5.

Performance Level Arguments
    --elastic-pool     : The name or resource id of the elastic pool to move the database into.
    --service-objective: The name of the new service objective. If this is a standalone db service
                         objective and the db is currently in an elastic pool, then the db is
                         removed from the pool.

Resource Id Arguments
    --ids              : One or more resource IDs (space-delimited). If provided, no other 'Resource
                         Id' arguments should be specified.
    --name -n          : Name of the Azure SQL Database.
    --resource-group -g: Name of resource group. You can configure the default group using `az
                         configure --defaults group=<name>`.
    --server -s        : Name of the Azure SQL server.
```

jaredmoo commented:
# Syntax changes (part 4: elastic pools)

## az sql elastic-pool create

### Before

```
Command
    az sql elastic-pool create: Create an elastic pool.

Arguments
    --name -n           [Required]: The name of the elastic pool.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --db-dtu-max --db-max-dtu     : The maximum DTU any one database can consume.
    --db-dtu-min --db-min-dtu     : The minimum DTU all databases are guaranteed.
    --dtu                         : The total shared DTU for the database elastic pool.
    --edition                     : The edition of the elastic pool.
    --max-size --storage          : The max storage size of the elastic pool. If no unit is
                                    specified, defaults to megabytes (MB).
    --tags                        : Resource tags.
    --zone-redundant -z           : Specifies whether to enable zone redundancy for the elastic
                                    pool.  Allowed values: false, true.
```

### After

```

Command
    az sql elastic-pool create: Create an elastic pool.

Arguments
    --name -n                        [Required]: The name of the elastic pool.
    --resource-group -g              [Required]: Name of resource group. You can configure the
                                                 default group using `az configure --defaults
                                                 group=<name>`.
    --server -s                      [Required]: Name of the Azure SQL server.
    --db-dtu-max --db-max-capacity --db-max-dtu: The maximum capacity any one database can consume.
    --db-dtu-min --db-min-capacity --db-min-dtu: The minimum capacity all databases are guaranteed.
    --license-type                             : The license type to apply for this elastic pool.
                                                 Allowed values: BasePrice, LicenseIncluded.
    --max-size --storage                       : The max storage size. If no unit is specified,
                                                 defaults to bytes (B).
    --tags                                     : Resource tags.
    --zone-redundant -z                        : Specifies whether to enable zone redundancy.
                                                 Allowed values: false, true.

Performance Level (components) Arguments
    --capacity --dtu -c                        : The capacity component of the sku in integer number
                                                 of DTUs or vcores.
    --edition --tier -e                        : The edition component of the sku. Allowed values
                                                 include: Basic, Standard, Premium, GeneralPurpose,
                                                 BusinessCritical.
    --family -f                                : The compute generation component of the sku (for
                                                 vcore skus only). Allowed values include: Gen4,
                                                 Gen5.
```

## az sql elastic-pool update

### Before

```
Command
    az sql elastic-pool update: Update an elastic pool.

Arguments
    --db-dtu-max --db-max-dtu: The maximum DTU any one database can consume.
    --db-dtu-min --db-min-dtu: The minimum DTU all databases are guaranteed.
    --dtu                    : TThe total shared DTU for the elastic eool.
    --max-size --storage     : Storage limit for the elastic pool in MB.
    --zone-redundant -z      : Specifies whether to enable zone redundancy for the elastic pool.
                               Allowed values: false, true.

Generic Update Arguments
    --add                    : Add an object to a list of objects by specifying a path and key value
                               pairs.  Example: --add property.listProperty <key=value, string or
                               JSON string>.
    --remove                 : Remove a property or an element from a list.  Example: --remove
                               property.list <indexToRemove> OR --remove propertyToRemove.
    --set                    : Update an object by specifying a property path and value to set.
                               Example: --set property1.property2=<value>.

Resource Id Arguments
    --ids                    : One or more resource IDs (space-delimited). If provided, no other
                               'Resource Id' arguments should be specified.
    --name -n                : The name of the elastic pool.
    --resource-group -g      : Name of resource group. You can configure the default group using `az
                               configure --defaults group=<name>`.
    --server -s              : Name of the Azure SQL server.
```

### After

```
Command
    az sql elastic-pool update: Update an elastic pool.

Arguments
    --db-dtu-max --db-max-capacity --db-max-dtu
    --db-dtu-min --db-min-capacity --db-min-dtu
    --max-size --storage                       : The max storage size. If no unit is specified,
                                                 defaults to bytes (B).
    --zone-redundant -z                        : Specifies whether to enable zone redundancy.
                                                 Allowed values: false, true.

Generic Update Arguments
    --add                                      : Add an object to a list of objects by specifying a
                                                 path and key value pairs.  Example: --add
                                                 property.listProperty <key=value, string or JSON
                                                 string>.
    --remove                                   : Remove a property or an element from a list.
                                                 Example: --remove property.list <indexToRemove> OR
                                                 --remove propertyToRemove.
    --set                                      : Update an object by specifying a property path and
                                                 value to set.  Example: --set
                                                 property1.property2=<value>.

Performance Level (components) Arguments
    --capacity --dtu -c                        : The capacity component of the sku in integer number
                                                 of DTUs or vcores.
    --edition --tier -e                        : The edition component of the sku. Allowed values
                                                 include: Basic, Standard, Premium, GeneralPurpose,
                                                 BusinessCritical.
    --family -f                                : The compute generation component of the sku (for
                                                 vcore skus only). Allowed values include: Gen4,
                                                 Gen5.

Resource Id Arguments
    --ids                                      : One or more resource IDs (space-delimited). If
                                                 provided, no other 'Resource Id' arguments should
                                                 be specified.
    --name -n                                  : The name of the elastic pool.
    --resource-group -g                        : Name of resource group. You can configure the
                                                 default group using `az configure --defaults
                                                 group=<name>`.
    --server -s                                : Name of the Azure SQL server.
```

## az sql elastic-pool list-editions

### Before

```
Command
    az sql elastic-pool list-editions: List elastic pool editions available for the active
    subscription.
        Also includes available pool DTU settings, storage limits, and per database settings. In
        order to reduce verbosity, additional storage limits and per database settings are hidden by
        default.

Arguments
    --location -l [Required]: Location. You can configure the default location using `az configure
                              --defaults location=<location>`.
    --show-details -d       : List of additional details to include in output.  Allowed values: db-
                              max-dtu, db-max-size, db-min-dtu, max-size.

Search Arguments
    --dtu                   : Elastic pool DTU limit to search for. If unspecified, all DTU limits
                              are shown.
    --edition               : Edition to search for. If unspecified, all editions are shown.
```

### After

```
Command
    az sql elastic-pool list-editions: List elastic pool editions available for the active
    subscription.
        Also includes available pool DTU settings, storage limits, and per database settings. In
        order to reduce verbosity, additional storage limits and per database settings are hidden by
        default.

Arguments
    --location -l [Required]: Location. You can configure the default location using `az configure
                              --defaults location=<location>`.
    --available -a          : If specified, show only results that are available in the specified
                              region.
    --show-details -d       : List of additional details to include in output.  Allowed values: db-
                              max-dtu, db-max-size, db-min-dtu, max-size.

Search Arguments
    --dtu                   : Number of DTUs to search for. If unspecified, all DTU sizes are shown.
    --edition --tier -e     : Edition to search for. If unspecified, all editions are shown.
    --vcores                : Number of vcores to search for. If unspecified, all vcore sizes are
                              shown.

```

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py:
```diff
@@ -615,20 +945,31 @@ def db_update(
 
-    # Null out edition so that edition gets chosen automatically by choice of SLO/pool
-    instance.edition = None
+    # TODO Temporary workaround for elastic pool sku name issue
```
I need to check if the fix for this is deployed to production yet.

jaredmoo commented:
Travis CI is seeing this error only on Python 2.7. It's not clear to me what the actual error was, and it doesn't seem related to anything I did.

https://travis-ci.org/Azure/azure-cli/jobs/376021771#L3811
```
======================================================================
ERROR: test_sql_elastic_pool_operation_mgmt (azure.cli.command_modules.sql.tests.latest.test_sql_commands.SqlElasticPoolOperationMgmtScenarioTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/azure_devtools/scenario_tests/preparers.py", line 52, in _preparer_wrapper
    fn(test_class_instance, **kwargs)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/azure_devtools/scenario_tests/preparers.py", line 41, in _preparer_wrapper
    **kwargs
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py", line 52, in create_resource
    execute(TestCli(), template.format(self.location, group, name, self.admin_user, self.admin_password))
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/azure/cli/testsdk/base.py", line 194, in __init__
    self._in_process_execute(cli_ctx, command, expect_failure=expect_failure)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/azure/cli/testsdk/base.py", line 245, in _in_process_execute
    self.exit_code = cli_ctx.invoke(shlex.split(command), out_file=stdout_buf) or 0
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/knack/cli.py", line 197, in invoke
    cmd_result = self.invocation.execute(args)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/azure/cli/core/commands/__init__.py", line 271, in execute
    parsed_args = self.parser.parse_args(args)
  File "/opt/python/2.7.14/lib/python2.7/argparse.py", line 1704, in parse_args
    self.error(msg % ' '.join(argv))
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/azure/cli/core/parser.py", line 124, in error
    self.exit(2)
  File "/opt/python/2.7.14/lib/python2.7/argparse.py", line 2362, in exit
    _sys.exit(status)
SystemExit: 2
-------------------- >> begin captured logging << --------------------
azure.cli.testsdk: INFO: Command "az group create --location southeastasia --name clitest.rg000001 --tag date=2018-05-07T18:41:28Z product=azurecli cause=automation" => 0. 
--------------------- >> end captured logging << ---------------------
```

Also seeing slow module load, not sure how to investigate. https://travis-ci.org/Azure/azure-cli/jobs/376021775#L1154
```
FAILED MODULES
Module                    Average    Threshold        Stdev                    Values
ALL                           365          300            7     [359.0, 375.0, 360.0]
sql                           232           30            3     [229.0, 237.0, 231.0]

```

jaredmoo commented:
I found how to add transforms that work with LRO. This will make backcompat much better. Working on it now.

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_format.py:
```diff
@@ -0,0 +1,254 @@
+    return str((b // unit[0])) + unit[1]
+
+
+class LongRunningOperationResultTransform(LongRunningOperation):  # pylint: disable=too-few-public-methods
```
You guys should consider adding this into CLI core. Or just fix this code so that it applies transform / LRO in the opposite order: https://github.com/azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/commands/__init__.py#L312

jaredmoo commented:
Phew, all checks passed. This is now ready for review please :)

jaredmoo commented:
Which param is missing help? It looks good to me. Yes the 3 long names is bad, partially explained in source code:
```
        # --db-dtu-max and --db-dtu-min were the original param names, which is consistent with the
        # 2014-04-01 REST API.
        # --db-max-dtu and --db-min-dtu are aliases which are consistent with the `sql elastic-pool
        # list-editions --show-details db-max-dtu db-min-dtu` parameter values. These are more
        # consistent with other az sql commands, but the original can't be removed due to
        # compatibility.
```

`--db-max-capacity` is the new 3rd name since the same REST API property ('capacity') can be measured in either DTU or VCores depending on the edition. If possible I would like to hide (from help) or deprecate one of the older options.

tjprescott commented:
@jaredmoo we'll be adding deprecation support for commands, groups, parameters and options soon, so that sounds like a plan.  Looks good for now.

tjprescott commented:
In your help dumps it looked like `elastic-pool update` was missing help for `--db-dtu-min/max`.

jaredmoo commented:
Thanks!
